framework/src/actions/CreateExecutionerAction.C

```diff
@@ -38,8 +40,14 @@ CreateExecutionerAction::act()
   _moose_object_pars.set<FEProblemBase *>("_fe_problem_base") = _problem.get();
   _moose_object_pars.set<FEProblem *>("_fe_problem") =
       (std::dynamic_pointer_cast<FEProblem>(_problem)).get();
+  _moose_object_pars.set<EigenProblem *>("_eigen_problem") =
+      (std::dynamic_pointer_cast<EigenProblem>(_problem)).get();
   std::shared_ptr<Executioner> executioner =
       _factory.create<Executioner>(_type, "Executioner", _moose_object_pars);
 
+  if (std::dynamic_pointer_cast<EigenProblem>(_problem).get() &&
+      !(std::dynamic_pointer_cast<EigenExecutioner>(executioner).get()))
+    mooseError("EigenProblem requires an EigenExecutioner or its derived subclass");
```

Comment:
Oh, I see, here's the check. We'll no need to perform these casts twice and in the wrong order. Save your cast results.